### DISCUSSION TITLE
New documentation

### DISCUSSION THREAD

**Discussion by franky47** [47ng/nuqs#409] [Ideas] [answered]
The playground started as a shell for e2e testing, but evolved into disorganised demos and showcases of what can be done with the library.

Also, the README is starting to become big and hard to digest, so a documentation overhaul would be welcome.

The objective is to keep a single deployable web app that does the following:
- Landing page with overview of features, and a simple demo to showcase sharing.
- Tutorial / Getting started with demos on replacing existing useState
- Recipes / how-to with concrete examples
- ~Internal pages dedicated to e2e testing, that can be unreachable (no need to link to them). Ideally, they should not be deployed on Vercel and only be used in CI.~ ->  The e2e test bench has been separated in #390, and will be kept lean to build quickly in CI.
- Serve as an URL for errors with additional context and remediation, to keep the bundle size small.

Maybe an API reference could be useful (currently covered by the README), especially an exhaustive list of parsers.

### Requirements & Specifications

- Be based on Next.js, to showcase the library (rules out Docusaurus)
- Work with the app router (rules out Nextra)
- Support local MDX files for content writing (allows co-location with embedded "use client" demos)
- Search with a simple integration story (Possible solutions: [cmdk](https://cmdk.paco.me/), [minisearch](https://github.com/lucaong/minisearch/), [Algolia DocSearch](https://docsearch.algolia.com/docs/what-is-docsearch/))
- Syntax highlighting with Shiki
- Styling with TailwindCSS + shadcn/ui for interactive components

Inspirations:
- https://diataxis.fr/
- https://github.com/t3-oss/t3-env/tree/main/docs/
- https://vitepress.dev
- https://github.com/fuma-nama/next-docs

### Installation

- Package manager instructions (npm, yarn, pnpm, bun)
- What version should I use (interactive?)
- Pinning versions (NUQS-409)

### Basic usage

- Replacing React.useState
- Default parser is a string
- URL to value mapping (nulls)

### Going further

- Parsers
- Options
- Testing
- Server cache

### Advanced usage

- Debugging
- SEO
- Throttling server updates
- Batching & awaiting URL updates
- Making your own parsers (w/ lossy caveat)

## Recipes (playground)

- Pagination
- Search 
- Table filters & sorting options
- Sliders to control UI (colors?)

## Reference

- useQueryState
- useQueryStates
- Parsers
- createSearchParamsCache
- subscribeToQueryUpdates _(this may be deprecated with WHS)_

**Comment by CogitoBurrito** [comment]
Hi, @franky47. How can I help?

**Reply by franky47** [reply]
Thanks for your help! I'm mainly at the planning stage, docs are not something to rush, especially wrt/ content.

The main area where I don't have much visibility at the moment is on the **presentation** side. The "requirements" section in the OP gives an overview of what would be ideal, but I don't know of a tool that currently does it all, especially for the app router. [`fuma-nama/next-docs`](https://github.com/fuma-nama/next-docs) seemed to tick a lot of boxes, but it's still in its infancy and I wasn't able to get it running locally.

Ideally, I'd want something low-maintenance, where we can focus on delivering good content. The moto here is:
```
Content > Presentation > Details
```

The other aspect I want to focus on in the docs is **interactivity**. The playground started as a showcase of what the library can do, but those turned out to become mere sandboxes for a quick test, and were left abandoned. Getting good, practical use-cases, that make sense to play with, placed alongside the documentation of a feature they showcase, is super important for the user experience. I've seen this a lot with the interactive articles on my blog.

Do you know of other tools that may help here?

**Reply by CogitoBurrito** [reply]
There are couple options for us. For interactivity we can use [Sandpack](https://sandpack.codesandbox.io/) or [WebContainer](https://webcontainers.io/)

>Edit looks like Docusaurus doesn't work with sandpack. We have to build the doc site using next.js

**Reply by CogitoBurrito** [reply]
Another option could be using Docusaurus for doc and stackblitz for interactivity, this will save a lot of time from building the site.

**Reply by franky47** [reply]
> Edit looks like Docusaurus doesn't work with sandpack. We have to build the doc site using Next.js

In this case we could go with Nextra, which hopefully will support the app router one day. Also I really much prefer the styles of Nextra Vs Docusaurus.

Regarding moving interactive parts to an external sandbox service, what would be the advantage of StackBlitz over CodeSandbox? I've had a lot of issues with docs sites embedding CodeSandboxes, where they wouldn't run with privacy-preserving settings and extensions.

Moreover, using those services, even if they offer a free tier now, could incur a cost in the future, at which point integrating examples in the docs site will be necessary.

But it's a great idea to get started quickly, and it allows others to remix and play with the examples in a way that wouldn't be possible if they were baked in the docs site.

**Reply by CogitoBurrito** [reply]
When you are running a dev environment: 
CodeSandbox is running a micro vm on their server, the web editor is only the frontend. 

StackBlitz, everything runs in the browser. 

> Update: StackBlitz cannot run nextjs 14 atm(waiting for nextjs team to update swc-wasm-nodejs). Codesanbox is the only option left.

BTW, [rspress](https://github.com/web-infra-dev/rspress) is a very good option as well(my personal favorite).
<img width="1503" alt="image" src="https://github.com/47ng/next-usequerystate/assets/12595973/f9472c99-aeda-47dd-b286-708e48de54c3">

**Reply by franky47** [reply]
@Talent30 I gave next-docs another try today and everything ran fine, even importing client and server components in MDX, so we should be good for embedded interactive demos.

Sandboxes will still make sense as a starting point for users to play with, by not requiring to clone/bootstrap the docs website.

I'll have to give Sandpack a try, but I'm wary of a sandbox trying to emulate browser APIs (history) and how a sandboxed Next.js might work around that, especially for server-oriented features (non-shallow updates, server cache).

**Reply by CogitoBurrito** [reply]
I have tried sandpack. Sadly it doesn’t work with newer version of nextjs. Their template is still running next12.

**Reply by CogitoBurrito** [reply]
stackblitz is working with the latest nextjs now. https://github.com/vercel/next.js/commit/bb25edfdd4964994176085f26a9f38434ad66b58
https://stackblitz.com/edit/stackblitz-starters-cyeg8a?file=README.md

**Reply by franky47** [reply]
CodeSandbox is kind of unreliable regarding Next.js routing:

https://github.com/47ng/next-usequerystate/issues/430#issuecomment-1853241182

**Reply by franky47** [reply]
> stackblitz is working with the latest nextjs now. https://github.com/vercel/next.js/commit/bb25edfdd4964994176085f26a9f38434ad66b58
https://stackblitz.com/edit/stackblitz-starters-cyeg8a?file=README.md

I just tried it and it doesn't update the query string in the browser-in-browser URL, but it does if you open the preview in a new tab.

**Comment by franky47** [comment, marked as answer]
Added PR #433 with a next-docs app containing a copy of the README as an example.

Looks great out of the box, though images don't seem to be supported due to the Next.js `<Image>` component requiring image dimension props. I have managed to fix that on my personal website though, so probably can backport there.

**Reply by CogitoBurrito** [reply]
Have you decided which way are you going to use for the interactive demo?

**Reply by franky47** [reply]
Not yet, CodeSandbox is unreliable, and StackBlitz doesn't show the query in their preview browser-in-browser UI.

I'm looking at GitHub Codespaces too, though that would be more beneficial for issue reproduction.

The plan for the docs PR is to merge the playground and docs apps into a single app, then merge onto `next` to allow content contributions.

**Reply by franky47** [reply]
Docs have been merged and are now deployed at [nuqs.47ng.com](https://nuqs.47ng.com).

There's still some work on:
- Splitting the readme into sections and pages (according to the outline above)
- Presentation on the landing page ("used by" section, footer)
- Fixing the playground pages styles (the tailwind CSS reset broke default browser styles)
- Renaming `next-usequerystate` into `nuqs` everywhere relevant
